README.md
# throttler  [![Build Status](https://travis-ci.org/daniel-gil/throttler.svg?branch=master)](https://travis-ci.org/daniel-gil/throttler) [![Coverage Status](https://coveralls.io/repos/github/daniel-gil/throttler/badge.svg?branch=master)](https://coveralls.io/github/daniel-gil/throttler?branch=master) [![Report card](https://goreportcard.com/badge/github.com/daniel-gil/throttler)](https://goreportcard.com/report/github.com/daniel-gil/throttler) ![Project status](https://img.shields.io/badge/version-0.1.1-green.svg)  ![Project dependencies](https://img.shields.io/badge/dependencies-none-green.svg) [![License: MIT](https://img.shields.io/badge/License-MIT-yellow.svg)](https://opensource.org/licenses/MIT) [![GoDoc](https://godoc.org/github.com/daniel-gil/throttler?status.svg)](https://godoc.org/github.com/daniel-gil/throttler)

Provides a throttle request channel for Go that controls request rate limit in order to prevent exceeding a predefined API quota.

## Installation

```sh
go get github.com/daniel-gil/throttler
```

## Motivation

Why we need to control the request rate?

Because some APIs limit the maximal number of calls that a client can request in order to control the received traffic. If this rate limit is overtaken then the sender will receive an HTTP Error from the server, for example a `403 Developer Over Rate`.

To avoid this situation the `throttler` package uses a mechanism to regulate the request flow to the server based on the leaky bucket algorithm where the bucket is represented by a buffered channel.

## Documentation

The throttler working flow has the following steps:

* Initialization: `Rate` creation passed to the throttler constructor `New`.
* Start the service with `Run`.
* Add new requests to be processed with `Queue`.

API documentation is available on [godoc.org][doc].

### Rate

`Rate` returns the minimal rate duration to communicate with a server that can not be overpassed. It is calculated as

```Rate Duration = period + guardTime```

where `period` represents the frequency that we should send the requests and `guardTime` is an extra time to wait beetween two consecutive calls.

The available `Rate` constructors are `NewRateByCallsPerSecond`, `NewRateByCallsPerMinute` or `NewRateByCallsPerHour`.

### New

The throttler constructor `New` is the responsible for initializing the requests channel and configuring the listener for this channel based on the `Rate` passed.

### Run

It starts a mechanism called `listener` in a new goroutine which controls that the requests received from the requests channel are fulfilled at the proper time respecting the `Rate` limits.

### Queue

The `Queue` function queues a new `throttler.Request` (which contains an `http.Request`) to the shared requests channel and blocks the thread until the `listener` decides that the request can be processed. When this happens, the function `fulfill` is called which internally calls the `http.Client.Do(http.Request)`. Finally the `Queue` function returns an `http.Response`.


## Usage

To use this package the first thing we have to do is create a `Rate` instance using any of the available constructors:

```go

rate, err := throttler.NewRateByCallsPerSecond(maxCallsPerSecond, guardTime)
rate, err := throttler.NewRateByCallsPerMinute(maxCallsPerMin, guardTime)
rate, err := throttler.NewRateByCallsPerHour(maxCallsPerHour, guardTime)

```

where `maxCallsPerSecond`, `maxCallsPerMin` and `maxCallsPerHour` are integers that define the maximal number of requests allowed to send to the client, and `guardTime` is an extra time to wait beetween two consecutive calls.

Then we create an instance of a throttler passing the rate:

```go

t, err := throttler.New(rate, requestChannelCapacity, client, verbose)

```

where `requestChannelCapacity` is the capacity of the channel that will contains all the queued requests, `client` is an optional parameter of type `*http.Client` and  `verbose` is a boolean that displays debug information in the standard output if `true`.

If the client param is nil is set to `http.DefaultClient`. The throttler allows to customize the `http.Client` because sometimes it is desired to specify timeouts, configure a redirect policy, proxies or simply to be used with Google App Engine.

Once we have an instance of the throttler we can activate the listener which will be waiting for new requests from the requests channel:
```go

t.Run()

```

After this we are ready to start queuing new requests to the channel:

```go

res, err := t.Queue(ctx, name, req, timeout)

```

where `ctx` is the context (used for cancellation propagation), `name` is an optional field used just for verbose, `req` is the request of type `http.Request` and `timeout` is the request timeout of type `time.Duration`.

### Concurrent Requests
In some situations we need to send multiple calls in parallel and we would like to avoid blocking the thread, for this case
we can achieve this by using goroutines. If we encapsulate the `Queue` call into a function like this:

```go

func handleRequest(ctx context.Context, name string, req *http.Request, timeout time.Duration) *http.Response {
    cres, err := t.Queue(ctx, name, req, timeout)
    if err != nil {
        log.Fatalf("unable to queue the request: %v", err)
    }
    return res
}

```

then we could call the `handleRequest` in a new goroutine and send as many requests as we need.
Later we can wait for the responses using a select case statement which controls if a global timeout has expired 
to stop waiting:

```go

c := make(chan *http.Response)
for i := 0; i < numRequests; i++ {
    name := "Task " + strconv.Itoa(i)
    go func() {
        c <- handleRequest(ctx, name, req, reqTimeout)
    }()
}
timeout := time.After(globalTimeout)
for i := 0; i < numRequests; i++ {
    select {
    case result := <-c:
        processResponse(i, result)
    case <-timeout:
        fmt.Printf("timed out")
        return
    }
}

```


## Example

There is a complete example under the `/example/` folder in the source code that shows how the package is used. The `/example/main.go` file accepts the following flags:

* **numReq**: Number of requests to call in parallel (default: 10).
* **reqChanCap**: Capacity of the requests channel (default: 10).
* **maxCallsPerSec**: Maximal number of calls per second (default: 2).
* **guardTimeInMs**: Extra time in miliseconds to wait between two consecutive calls (default: 50).
* **reqTimeoutInMs**: Request timeout in miliseconds (default: 10000, it is 10 seconds).
* **globalTimeoutInMs**: Global timeout in miliseconds for sending all the requests (default: 30000, it is 30 seconds).
* **verbose**: If true prints information about the requests fulfilled by the throttler: name, timestamp, order (default: true).


Output:

```sh

$ go run ./example/main.go
Throttler started
10 request(s) pending to be processed at Rate = (1 call / 550ms).

[2018-03-16 11:38:20.797001121 +0100 CET m=+0.554512486] got ticket; Fulfilling Request [Task 0]
[2018-03-16 11:38:20.797321571 +0100 CET m=+0.554832931] Request fulfilled [Task 0]
[2018-03-16 11:38:21.348416765 +0100 CET m=+1.105919716] got ticket; Fulfilling Request [Task 2]
[2018-03-16 11:38:21.348537267 +0100 CET m=+1.106040216] Request fulfilled [Task 2]
[2018-03-16 11:38:21.898448991 +0100 CET m=+1.655943551] got ticket; Fulfilling Request [Task 1]
[2018-03-16 11:38:21.898548044 +0100 CET m=+1.656042602] Request fulfilled [Task 1]
[2018-03-16 11:38:22.443900051 +0100 CET m=+2.201386288] got ticket; Fulfilling Request [Task 5]
[2018-03-16 11:38:22.443991627 +0100 CET m=+2.201477863] Request fulfilled [Task 5]
[2018-03-16 11:38:22.996845058 +0100 CET m=+2.754322858] got ticket; Fulfilling Request [Task 3]
[2018-03-16 11:38:22.996902877 +0100 CET m=+2.754380676] Request fulfilled [Task 3]
[2018-03-16 11:38:23.547879646 +0100 CET m=+3.305349039] got ticket; Fulfilling Request [Task 7]
[2018-03-16 11:38:23.547948586 +0100 CET m=+3.305417978] Request fulfilled [Task 7]
[2018-03-16 11:38:24.093824873 +0100 CET m=+3.851285936] got ticket; Fulfilling Request [Task 6]
[2018-03-16 11:38:24.093901681 +0100 CET m=+3.851362743] Request fulfilled [Task 6]
[2018-03-16 11:38:24.645247933 +0100 CET m=+4.402700583] got ticket; Fulfilling Request [Task 8]
[2018-03-16 11:38:24.645312707 +0100 CET m=+4.402765356] Request fulfilled [Task 8]
[2018-03-16 11:38:25.195285184 +0100 CET m=+4.952729442] got ticket; Fulfilling Request [Task 9]
[2018-03-16 11:38:25.195345115 +0100 CET m=+4.952789372] Request fulfilled [Task 9]
[2018-03-16 11:38:25.747991251 +0100 CET m=+5.505427075] got ticket; Fulfilling Request [Task 4]
[2018-03-16 11:38:25.748063331 +0100 CET m=+5.505499154] Request fulfilled [Task 4]

Elapsed time: 5.821642877s

```

## Tests

### client_test.go

Contains a test case for testing the `send` function.


### export_test.go

Contains some alias to be able to access privave functions just for testing.

### fulfiller_test.go

Contains a test case for testing the `fulfill` function.

### listener_test.go

Contains a test case for testing the `listen` function.


### handler_test

The file `handler_test.go` contains some test cases for testing the functions `NewHandler`, `SetClient`, `Run` and `Queue`.

### rate_test.go

Contains test cases for testing the rate functions `NewRateByCallsPerSecond`, `NewRateByCallsPerMinute`, `NewRateByCallsPerHour` and `CalculateRate`.

### throttler_test.go

Contains test cases for testing the functions `New`, `Rate`, `Run` and `Queue`.


### Run all tests

Use the following command to run all the tests:

```sh

go test -v -race

```

Output:
```sh

$ go test -race
[2018-03-16 11:37:45.443431875 +0100 CET m=+1.575356426] got ticket; Fulfilling Request [Positive TC]
[2018-03-16 11:37:45.444484452 +0100 CET m=+1.576408987] Request fulfilled [Positive TC]
PASS
[2018-03-16 11:37:46.785455926 +0100 CET m=+2.917360001] got ticket; Fulfilling Request [Negative TC: force timeout in Queue]
[2018-03-16 11:37:46.785917845 +0100 CET m=+2.917821913] Request fulfilled [Negative TC: force timeout in Queue]
ok      github.com/daniel-gil/throttler  3.389s

```

## Limitations

* If we have multiple application instances could be possible to overtake the rate limit because the instances do not share the requests channel.
* The fulfill function always calls `http.Client.Do(http.Request)`.


## References

* [Rate Limiting](https://github.com/golang/go/wiki/RateLimiting)
* [Rate Limiting Service Calls in Go](https://medium.com/@KevinHoffman/rate-limiting-service-calls-in-go-3771c6b7c146) by Kevin Hoffman
* [Mocking in Go (dotGo 2014)](https://youtu.be/2_FMbcQJg0c) by Gabriel Aszalos


## License

This project is under the [MIT License][mit].

[mit]: https://github.com/daniel-gil/throttler/blob/master/LICENSE
[doc]: https://godoc.org/github.com/daniel-gil/throttler

CHANGELOG.md
# Changelog

## [0.1.1] - 2019-04-12
### Changed
- Updated documentation
- Fixed references

## [0.1.0] - 2018-03-16
### Changed
- Convert private methods into structs with an interface (fulfiller, client and listener) that can be injected, it makes easier testing all parts of the code.

## [0.0.3] - 2018-03-07
### Update
- Updated Readme.md documentation.

## [0.0.2] - 2018-03-07
### Fixed
- Corrected references to packages.
- Display correctly Readme.md badges.

## [0.0.1] - 2018-03-06

Initial release


rate_test.go
package throttler_test

import (
	"testing"
	"time"

	"github.com/daniel-gil/throttler"
)

func TestNewRateByCallsPerSecond(t *testing.T) {
	tt := []struct {
		name                 string
		maxCallsPerSecond    int
		guardTime            time.Duration
		expectedRateDuration time.Duration
		errMsg               string
	}{
		{"Positive TC", 2, duration50ms, duration550ms, ""},
		{"Negative TC: maxCallsPerSecond zero", 0, duration50ms, 0, "maxCalls must be greater than zero"},
		{"Negative TC: maxCallsPerSecond", -2, duration50ms, 0, "maxCalls must be greater than zero"},
		{"Negative TC: guardTime", 2, -duration50ms, 0, "guardTime must be greater or equal than zero"},
		{"Positive TC: guardTime zero", 2, 0, duration500ms, ""},
	}

	for _, tc := range tt {
		t.Run(tc.name, func(t *testing.T) {
			rate, err := throttler.NewRateByCallsPerSecond(tc.maxCallsPerSecond, tc.guardTime)
			if !checkError(tc.errMsg, err, t) {
				rateDuration := rate.CalculateRate()
				if rateDuration != tc.expectedRateDuration {
					t.Errorf("expected rate duration %v; got %v", tc.expectedRateDuration, rateDuration)
				}
			}
		})
	}
}

func TestNewRateByCallsPerMinute(t *testing.T) {
	tt := []struct {
		name                 string
		maxCallsPerMinute    int
		guardTime            time.Duration
		expectedRateDuration time.Duration
		errMsg               string
	}{
		{"Positive TC", 2, duration5s, duration35s, ""},
		{"Negative TC: maxCallsPerMinute zero", 0, duration50ms, 0, "maxCalls must be greater than zero"},
		{"Negative TC: maxCallsPerMinute", -2, duration50ms, 0, "maxCalls must be greater than zero"},
		{"Negative TC: guardTime", 2, -duration50ms, 0, "guardTime must be greater or equal than zero"},
		{"Positive TC: guardTime zero", 2, 0, duration30s, ""},
	}

	for _, tc := range tt {
		t.Run(tc.name, func(t *testing.T) {
			rate, err := throttler.NewRateByCallsPerMinute(tc.maxCallsPerMinute, tc.guardTime)
			if !checkError(tc.errMsg, err, t) {
				rateDuration := rate.CalculateRate()
				if rateDuration != tc.expectedRateDuration {
					t.Errorf("expected rate duration %v; got %v", tc.expectedRateDuration, rateDuration)
				}
			}
		})
	}
}

func TestNewRateByCallsPerHour(t *testing.T) {
	tt := []struct {
		name                 string
		maxCallsPerHour      int
		guardTime            time.Duration
		expectedRateDuration time.Duration
		errMsg               string
	}{
		{"Positive TC", 2, duration5min, duration35min, ""},
		{"Negative TC: maxCallsPerHour zero", 0, duration50ms, 0, "maxCalls must be greater than zero"},
		{"Negative TC: maxCallsPerHour", -2, duration50ms, 0, "maxCalls must be greater than zero"},
		{"Negative TC: guardTime", 2, -duration50ms, 0, "guardTime must be greater or equal than zero"},
		{"Positive TC: guardTime zero", 2, 0, duration30min, ""},
	}

	for _, tc := range tt {
		t.Run(tc.name, func(t *testing.T) {
			rate, err := throttler.NewRateByCallsPerHour(tc.maxCallsPerHour, tc.guardTime)
			if !checkError(tc.errMsg, err, t) {
				rateDuration := rate.CalculateRate()
				if rateDuration != tc.expectedRateDuration {
					t.Errorf("expected rate duration %v; got %v", tc.expectedRateDuration, rateDuration)
				}
			}
		})
	}
}

func TestCalculateRate(t *testing.T) {
	tt := []struct {
		name              string
		maxCallsPerSecond int
		guardTime         time.Duration
		expectedRate      time.Duration
	}{
		{"Positive TC", 2, duration50ms, 550 * time.Millisecond},
	}

	for _, tc := range tt {
		t.Run(tc.name, func(t *testing.T) {
			rate, err := throttler.NewRateByCallsPerSecond(tc.maxCallsPerSecond, duration50ms)
			if err != nil {
				t.Fatalf("unable to create a rate")
			}
			rateDuration := rate.CalculateRate()
			if rateDuration != tc.expectedRate {
				t.Errorf("unexpected rate duration, expected 550ms; got %v", rateDuration)
			}
		})
	}
}


throttler_test.go
package throttler_test

import (
	"context"
	"fmt"
	"net/http"
	"testing"
	"time"

	"github.com/daniel-gil/throttler"
)

var duration35min = 35 * time.Minute
var duration30min = 30 * time.Minute
var duration5min = 5 * time.Minute

var duration35s = 35 * time.Second
var duration30s = 30 * time.Second
var duration5s = 5 * time.Second
var duration550ms = 550 * time.Millisecond
var duration500ms = 500 * time.Millisecond
var duration50ms = 50 * time.Millisecond
var duration10s = 10 * time.Second
var duration1ns = 1 * time.Nanosecond

type MockRate struct {
	CalculateRateMock func() time.Duration
}

func (m *MockRate) CalculateRate() time.Duration {
	return m.CalculateRateMock()
}

func buildThrottler(rate throttler.Rate, maxCallsPerSecond int, guardTime time.Duration, requestChannelCapacity int, verbose bool, client *http.Client) (throttler.Limiter, error) {
	limiter, err := throttler.New(rate, requestChannelCapacity, client, verbose)
	if err != nil || limiter == nil {
		return nil, fmt.Errorf("unable to create a new throttler: %v", err)
	}
	return limiter, nil
}

func TestNew(t *testing.T) {
	tt := []struct {
		name              string
		maxCallsPerSecond int
		guardTime         time.Duration
		reqChanCapacity   int
		errMsg            string
	}{
		{"Positive TC", 2, duration50ms, 5, ""},
		{"Positive TC: reqChanCapacity zero", 2, duration50ms, 0, ""},
		{"Negative TC: reqChanCapacity negative", 2, duration50ms, -5, "reqChanCapacity must be greater than zero"},
		{"Negative TC: rate nil", 0, 0, 10, "rate can not be nil"},
		{"Negative TC: maxCalls zero", 0, 2, 10, "maxCalls must be greater than zero"},
		{"Negative TC: new listener with errors", 2, 10, 10, "maxCalls must be greater than zero"},
	}

	for _, tc := range tt {
		t.Run(tc.name, func(t *testing.T) {
			var rate throttler.Rate
			var err error
			var foundError = false
			if tc.name != "Negative TC: rate nil" {
				rate, err = throttler.NewRateByCallsPerSecond(tc.maxCallsPerSecond, tc.guardTime)
				if err != nil {
					if checkError(tc.errMsg, err, t) {
						foundError = true
					}

				}
			}
			if !foundError {
				var client *http.Client
				throttler, err := throttler.New(rate, tc.reqChanCapacity, client, false)
				if throttler == nil {
					checkError(tc.errMsg, err, t)
				}
			}
		})
	}
}

func TestRate(t *testing.T) {
	tt := []struct {
		name         string
		expectedRate time.Duration
	}{
		{"Positive TC", 5 * time.Second},
	}

	for _, tc := range tt {
		t.Run(tc.name, func(t *testing.T) {
			called := false
			mockRate := &MockRate{
				CalculateRateMock: func() time.Duration {
					called = true
					return tc.expectedRate
				},
			}
			throttler, err := buildThrottler(mockRate, 2, duration50ms, 10, false, nil)
			if err != nil {
				t.Fatalf("unable to create a throttler")
			}
			rateDuration := throttler.Rate()
			if !called {
				t.Fatalf("did not call CalculateRate")
			}
			if rateDuration != tc.expectedRate {
				t.Errorf("expected rate duration %v; got %v", tc.expectedRate, rateDuration)
			}
			throttler.Run()
		})
	}
}

func TestRun(t *testing.T) {
	tt := []struct {
		name         string
		expectedRate time.Duration
	}{
		{"Positive TC", 5 * time.Second},
	}

	for _, tc := range tt {
		t.Run(tc.name, func(t *testing.T) {
			called := false
			mockRate := &MockRate{
				CalculateRateMock: func() time.Duration {
					called = true
					return tc.expectedRate
				},
			}

			throttler, err := buildThrottler(mockRate, 2, duration50ms, 10, false, nil)
			if err != nil {
				t.Fatalf("unable to create a throttler")
			}
			rateDuration := throttler.Rate()
			if !called {
				t.Fatalf("did not call CalculateRate")
			}
			if rateDuration != tc.expectedRate {
				t.Errorf("expected rate duration %v; got %v", tc.expectedRate, rateDuration)
			}

		})
	}
}

func TestQueue(t *testing.T) {
	tt := []struct {
		name                string
		maxCallsPerSecond   int
		guardTime           time.Duration
		reqChanCapacity     int
		verbose             bool
		timeout             time.Duration
		startRequestHandler bool
		errMsg              string
	}{
		{"Positive TC", 2, duration50ms, 5, true, duration10s, true, ""},
		{"Negative TC: requestHandler not started", 2, duration50ms, 5, false, duration10s, false, "requestHandler has not been started"},
		{"Negative TC: force timeout in Queue", 2, duration50ms, 5, true, duration1ns, true, "context deadline exceeded"},
	}

	for _, tc := range tt {
		t.Run(tc.name, func(t *testing.T) {
			rate, err := throttler.NewRateByCallsPerSecond(tc.maxCallsPerSecond, tc.guardTime)
			if err != nil {
				t.Fatalf("unable to create a rate")
			}
			throttler, err := throttler.New(rate, tc.reqChanCapacity, nil, tc.verbose)

			if tc.startRequestHandler {
				throttler.Run()
			}
			if err != nil {
				t.Fatalf("unable to create throttler: %v", err)
			}
			ctx := context.Background()
			req, err := http.NewRequest("GET", "https://golang.org/", nil)
			if err != nil {
				t.Fatalf("unable to create request: %v", err)
			}
			res, err := throttler.Queue(ctx, tc.name, req, tc.timeout)
			if !checkError(tc.errMsg, err, t) {
				if res.StatusCode != http.StatusOK {
					t.Fatalf("expected http.StatusOK (200); got: %v", res.Status)
				}
			}
		})
	}
}

/*

func TestRun(t *testing.T) {
	tt := []struct {
		name              string
		maxCallsPerSecond int
		guardTime         time.Duration
		reqChanCapacity   int
		errMsg            string
	}{
		{"Positive TC", 2, duration50ms, 5, ""},
	}

	for _, tc := range tt {
		t.Run(tc.name, func(t *testing.T) {
			rate, err := throttler.NewRateByCallsPerSecond(tc.maxCallsPerSecond, tc.guardTime)
			if err != nil {
				t.Fatalf("unable to create a rate")
			}

			// TODO: could we test the function Run() without instanciating the NewHandler?
			handler, err := throttler.New(rate, tc.reqChanCapacity, false, nil)
			if err != nil {
				t.Fatalf("unable to create a handler")
			}

			handler.Run()
			// TODO: how can we check that this was set correctly: h.requestHandlerStarted = true (if it is not public)
		})
	}
}


*/

func checkError(errMsg string, err error, t *testing.T) bool {
	if err != nil {
		if errMsg == "" {
			// here the testcase didn't expect any error
			t.Errorf("unexpected error: %v", err)
		} else if errMsg != err.Error() {
			// here the testcase expected another error than the received
			t.Errorf("expected error message: %v; got: %v", errMsg, err.Error())
		}
		return true
	}
	return false
}


example/main.go
package main

import (
	"context"
	"flag"
	"fmt"
	"io/ioutil"
	"log"
	"net/http"
	"strconv"
	"time"

	"github.com/daniel-gil/throttler"
)

var t throttler.Limiter

func main() {
	start := time.Now()

	// read the flags
	numRequests := flag.Int("numReq", 10, "number of requests to call in parallel")
	reqChanCap := flag.Int("reqChanCap", 10, "capacity of the requests channel")
	maxCallsPerSecond := flag.Int("maxCallsPerSec", 2, "maximal number of calls per second")
	guardTimeInMs := flag.Int("guardTimeInMs", 50, "extra time to wait between two consecutive calls (in miliseconds)")
	reqTimeoutInMs := flag.Int("reqTimeoutInMs", 10000, "request timeout (in miliseconds)")
	globalTimeoutInMs := flag.Int("globalTimeoutInMs", 30000, "global timeout (in miliseconds) for sending all the requests")
	verbose := flag.Bool("verbose", true, "if true prints information about the requests fulfilled by the throttler handler (name, timestamp, order)")

	flag.Parse()
	guardTime := time.Duration(*guardTimeInMs) * time.Millisecond
	reqTimeout := time.Duration(*reqTimeoutInMs) * time.Millisecond
	globalTimeout := time.Duration(*globalTimeoutInMs) * time.Millisecond

	fmt.Println("Throttler started")

	// initialize the throttler
	t = initThrottler(*maxCallsPerSecond, guardTime, *reqChanCap, *verbose, globalTimeout)
	req := createRequest()
	ctx := context.Background()
	c := make(chan *http.Response)

	// Generate some requests and queue them to the requests channel to be processed
	// when it corresponds (according to the maxCallsPerSecond configuration)
	for i := 0; i < *numRequests; i++ {
		name := "Task " + strconv.Itoa(i)
		go func() {
			c <- handleRequest(ctx, name, req, reqTimeout)
		}()
	}
	fmt.Printf("%d request(s) pending to be processed at Rate = (1 call / %v).\n\n", *numRequests, t.Rate())

	// Wait for receiving the responses from the channel and process each of them.
	// If a time out occurs, break the for loop.
	timeout := time.After(globalTimeout)
	for i := 0; i < *numRequests; i++ {
		select {
		case result := <-c:
			processResponse(i, result)
		case <-timeout:
			fmt.Printf("timed out")
			return
		}
	}
	elapsed := time.Since(start)
	fmt.Printf("\nElapsed time: %v\n", elapsed)
}

// initThrottler creates a new instance of a throttler.Handler
func initThrottler(maxCallsPerSecond int, guardTime time.Duration, requestChannelCapacity int, verbose bool, globalTimeoutInMs time.Duration) throttler.Limiter {
	rate, err := throttler.NewRateByCallsPerSecond(maxCallsPerSecond, guardTime)
	if err != nil {
		log.Fatalf("unable to create a rate: %v", err)
	}
	client := &http.Client{Timeout: globalTimeoutInMs}
	throttler, err := throttler.New(rate, requestChannelCapacity, client, verbose)
	if err != nil || throttler == nil {
		log.Fatalf("unable to create a new throttler: %v", err)
	}
	throttler.Run()
	return throttler
}

// createRequest creates a request to a test API to retrieve the current datetime
func createRequest() *http.Request {
	url := "http://date.jsontest.com/"
	req, err := http.NewRequest("GET", url, nil)
	if err != nil {
		log.Fatalf("unable to create request: %v", err)
	}
	return req
}

// handleRequest queues the http.Request to the handler requests channel
func handleRequest(ctx context.Context, name string, req *http.Request, requestTimeout time.Duration) *http.Response {
	res, err := t.Queue(ctx, name, req, requestTimeout)
	if err != nil {
		log.Fatalf("unable to queue the request: %v", err)
	}
	return res
}

// processResponse extracts the body from the http.Response and print it to the standard output
func processResponse(index int, res *http.Response) {
	if res.StatusCode != http.StatusOK {
		log.Fatalf("http status code is not OK: %s", res.Status)
	}
	defer res.Body.Close()
	body, err := ioutil.ReadAll(res.Body)
	if err != nil {
		log.Fatalf("Error reading response: %s", err)
	}
	if len(body) == 0 {
		log.Fatalf("response with empty body. Status: %s", res.Status)
	}
}
